### PR TITLE
Add unique constraint to MicrosoftImagesModel

### PR DESCRIPTION
**Pull request by rtamalin**
Add a UniqueConstraint based upon the name and environment columns to
the MicrosoftImagesModel (microsoftimages) table definition.

Add a schema migration to upgrade the database to add the new unique
constraint to the microsoftimages.

Add a new class method to the PintBase class that returns the list of
defined unique constraints associated with a given table, or an empty
list if none are found.

Update the data_update.py IDENTITY_OVERRIDES mechanism to leverage the
new unique_constraints class method from PintBase when specifying an
override entry for the 'microsoftimages' table.

Report any duplicates found for the proposed unique constraint.
    
Any that are found will be reported as WARNING messages prior to the
attempt to add the new unique constraint.

Bump the version to 2.0.5

Closes: #97

### CODE REVIEW

File: pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+revision = '8c5fc3cd9b18'
+down_revision = 'e2bdb3a5b1b4'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # Add a new unique constraint to the microsoft images table based upon
+    # name and environment columns.
+    op.create_unique_constraint(None, 'microsoftimages', ['name', 'environment'])
```
Comment (bear454) [inline pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py:22]: Before implement the database constraint, the migration script should ensure no existing records violate the constraint. In the case that any do, they should be listed so they can be manually fixed.

Without such a prevalidation, applying the constraint may fail, and _at best_ one record with the problem will be returned.

Such prevalidation should be performed before adding any database constraint through migration.

Ref: https://suse.slack.com/archives/C02A8AJAKPU/p1635184795066100

**Review by bear454 — CHANGES_REQUESTED** [review]

File: pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py (hunk 1 above)
Comment (rtamalin) [reply pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py:22]: The Alembic/SQLAlchemy migration will fail if the table schema change fails, as can be seen below when I recreated the issue that was encountered after the initial primary key updates change, where duplicated `microsoftimages` entries were created:

```
% ../bin/schema_upgrade.sh --host pdm-dev --user snotty --password MasterSlobs --database postgres --debug upgrade
DEBUG:__main__:db_uri: postgresql://snotty:MasterSlobs@pdm-dev:5432/postgres
DEBUG:__main__:repository: /home/fergal/work/public-cloud-team/public-cloud-info-service/pint_server/pint_db_migrate
INFO:__main__:Creating version control
INFO:__main__:Upgrading schema
INFO:alembic.runtime.migration:Context impl PostgresqlImpl.
INFO:alembic.runtime.migration:Will assume transactional DDL.
INFO:alembic.runtime.migration:Running upgrade e2bdb3a5b1b4 -> 8c5fc3cd9b18, microsoftimages unique constraint
DEBUG:__main__:(psycopg2.errors.UniqueViolation) could not create unique index "microsoftimages_name_environment_key"
DETAIL:  Key (name, environment)=(suse-sles-15-sp2-chost-byos-v20210304-gen2, PublicAzure) is duplicated.

[SQL: ALTER TABLE microsoftimages ADD UNIQUE (name, environment)]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
Traceback (most recent call last):
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.UniqueViolation: could not create unique index "microsoftimages_name_environment_key"
DETAIL:  Key (name, environment)=(suse-sles-15-sp2-chost-byos-v20210304-gen2, PublicAzure) is duplicated.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/fergal/work/public-cloud-team/public-cloud-info-service/pint_server/schema_upgrade.py", line 106, in upgrade
    command.upgrade(get_alembic_config(
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/command.py", line 294, in upgrade
    script.run_env()
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/script/base.py", line 490, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/util/pyfiles.py", line 97, in load_python_file
    module = load_module_py(module_id, path)
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/util/compat.py", line 184, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/fergal/work/public-cloud-team/public-cloud-info-service/pint_server/pint_db_migrate/env.py", line 78, in <module>
    run_migrations_online()
  File "/home/fergal/work/public-cloud-team/public-cloud-info-service/pint_server/pint_db_migrate/env.py", line 72, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/runtime/environment.py", line 813, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/runtime/migration.py", line 561, in run_migrations
    step.migration_fn(**kw)
  File "/home/fergal/work/public-cloud-team/public-cloud-info-service/pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py", line 22, in upgrade
    op.create_unique_constraint(None, 'microsoftimages', ['name', 'environment'])
  File "<string>", line 8, in create_unique_constraint
  File "<string>", line 3, in create_unique_constraint
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/operations/ops.py", line 375, in create_unique_constraint
    return operations.invoke(op)
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/operations/base.py", line 354, in invoke
    return fn(self, operation)
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/operations/toimpl.py", line 154, in create_constraint
    operations.impl.add_constraint(
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/ddl/impl.py", line 263, in add_constraint
    self._exec(schema.AddConstraint(const))
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/alembic/ddl/impl.py", line 146, in _exec
    return conn.execute(construct, multiparams)
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1011, in execute
    return meth(self, multiparams, params)
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/sql/ddl.py", line 72, in _execute_on_connection
    return connection._execute_ddl(self, multiparams, params)
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1068, in _execute_ddl
    ret = self._execute_context(
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1316, in _execute_context
    self._handle_dbapi_exception(
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1510, in _handle_dbapi_exception
    util.raise_(
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "/home/fergal/.venvs/pint/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) could not create unique index "microsoftimages_name_environment_key"
DETAIL:  Key (name, environment)=(suse-sles-15-sp2-chost-byos-v20210304-gen2, PublicAzure) is duplicated.

[SQL: ALTER TABLE microsoftimages ADD UNIQUE (name, environment)]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
Failed to upgrade Pint database: (psycopg2.errors.UniqueViolation) could not create unique index "microsoftimages_name_environment_key"
DETAIL:  Key (name, environment)=(suse-sles-15-sp2-chost-byos-v20210304-gen2, PublicAzure) is duplicated.

[SQL: ALTER TABLE microsoftimages ADD UNIQUE (name, environment)]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
```

So the migration will automatically fail if duplicates that break the new unique constraint are found, reporting the first such duplicate that is encountered, so from that perspective the code is fully functional, and needs no changes.

Comment (rtamalin) [reply pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py:22]: So what you are asking for is the addition of extra functionality which checks for the existence of potential duplicates, and lists any found before failing the schema migration.

I'm happy to do so, but would like to confirm some details on how we report this information.

- Where do we want the output presented?
Is it OK to just report them the WARNING stream of the schema_upgrade.py?
Or would you prefer the ERROR stream?
Or somewhere else?

- Output Format?
Do we want it reported as just the (`name`, `environment`) pairings for which duplicates exist?
Or do you want the full table entries for those duplicates? Or just the surrogate `id` key values associated with the duplicated entries?

Comment (rjschwei) [reply pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py:22]: IMHO we should check the input data and fix it there. The XML is still the truth. If we report it during this migration we still have to fix the issue in the XML. So first let's check the XML and if there are no issues we can skip the additional changes.

**Review by rjschwei — APPROVED** [review]

File: pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py (hunk 1 above)
Comment (smarlowucf) [reply pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py:22]: But if duplicates already made their way into the prod db then xml fixes won't fix the db. The XML, the import script or the db could be wrong. We would want to validate all three. I don't see anything in DB though:

```
pint=# SELECT name, environment FROM microsoftimages GROUP BY name, environment HAVING count(*) > 1;
 name | environment 
------+-------------
(0 rows)
```

Comment (rtamalin) [reply pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py:22]: @rjschwei @smarlowucf I "recreated" the problem scenario for the microsoftimages table (that has already been fixed!) to demonstrate that the SQLAlchemy/Alembic schema migration will automatically detect and complain about the presence of duplicates, failing the migration.

There is no problem in the current XML data, or the current data_update.py, or the current production DB, so we are good in that regard.

My goal with the about failure report was purely to demonstrate that if you had a DB with duplicated `microsoftimages` entries, like we ended up with previously, and tried to perform the proposed schema migration in this PR, the Alembic migration handling would automatically detect and fail, which is the desired behaviour for such a case.

This was in response to @bear454 requesting changes because the proposed schema migration code doesn't explicitly check for such duplicates before attempting to add the new unique constraint, where my intent is to demonstrate that we don't necessarily need to perform an explicit just to get the failure if the current `microsoftimages` table data isn't compatible with the new constraint.

The second part of @bear454 's request for changes is that he feels the schema migration scripting should report any potential duplicates that may exist, which is technically additional functionality over and above the normal schema migration handling, hence my request for clarification on how the duplicates should be reported.

Comment (rjschwei) [reply pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py:22]: IMHO we are done here.

- We have no duplicates
- Going forward we cannot introduce duplicates

As such I do not understand why we would need more test cases for duplicate entries.

Comment (bear454) [reply pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py:22]: As a design pattern, we do not want migrations that (a) add new constraints without first (b) providing a full list of any records that fail the constraint. That we internally at SUSE do not have a specific known issue in this migration should not be construed as solving this problem for any other potential user of this open source project, nor should we engage in a pattern that leads to intentional frustration.

As Fergal showed in the example test, a failed migration will return 1 and exactly 1 record that fails the constraint. If there are in fact 100 errors (or 1K, or 1M...), running a migration that applies a constraint will only show 1 failure at a time.

Having been through this exact scenario on other software service projects, I can this is an exceptionally frustrating problem, but thankfully it has a clear and evident pattern of solution: check first and report any errors, then proceed with migration if and only if there are no errors.

Comment (rtamalin) [reply pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py:22]: So I've been playing around within the migration script "space" and come up with a process to allow me to do the duplicates detection, using an embedded SQL statement; there are limits on what can be done using Alembic's `operations` object vs normal SQLAlchemy or SQLAlchemy ORM mechanisms, specifically when you just want to query the existing tables rather than update the content of the existing tables.

What I've currently prototyped yields logged output like the following:

```
% ../bin/schema_upgrade.sh --host pdm-dev --user snotty --password MasterSlobs --database postgres --debug upgrade
DEBUG:__main__:db_uri: postgresql://snotty:MasterSlobs@pdm-dev:5432/postgres
DEBUG:__main__:repository: /home/fergal/work/public-cloud-team/public-cloud-info-service/pint_server/pint_db_migrate
INFO:__main__:Creating version control
INFO:__main__:Upgrading schema
INFO:alembic.runtime.migration:Context impl PostgresqlImpl.
INFO:alembic.runtime.migration:Will assume transactional DDL.
INFO:alembic.runtime.migration:Running upgrade e2bdb3a5b1b4 -> 8c5fc3cd9b18, microsoftimages unique constraint
WARNING:8c5fc3cd9b18_microsoftimages_unique_constraint.py:Duplicates detected in microsoftimages:
WARNING:8c5fc3cd9b18_microsoftimages_unique_constraint.py:    {'name': '38dfb8618d9b41a3a48b1fe7549ae321__suse-sles-15-sp2-chost-byos-v20210304', 'environment': 'Mooncake'}
WARNING:8c5fc3cd9b18_microsoftimages_unique_constraint.py:    {'name': 'suse-sles-15-sp2-chost-byos-v20210304-gen2', 'environment': 'Mooncake'}
WARNING:8c5fc3cd9b18_microsoftimages_unique_constraint.py:    {'name': '5c9ba39cec434780938dba0f6ea3126d__suse-sles-15-sp2-chost-byos-v20210304', 'environment': 'Fairfax'}
WARNING:8c5fc3cd9b18_microsoftimages_unique_constraint.py:    {'name': 'suse-sles-15-sp2-chost-byos-v20210304-gen2', 'environment': 'Fairfax'}
WARNING:8c5fc3cd9b18_microsoftimages_unique_constraint.py:    {'name': 'suse-sles-15-sp2-chost-byos-v20210304-gen2', 'environment': 'PublicAzure'}
WARNING:8c5fc3cd9b18_microsoftimages_unique_constraint.py:    {'name': '960f9f43376d45d38de1615dd9d8e96a__suse-sles-15-sp2-chost-byos-v20210304', 'environment': 'Blackforest'}
WARNING:8c5fc3cd9b18_microsoftimages_unique_constraint.py:    {'name': '021d1b90c82943ec959408cff8e26c37__suse-sles-15-sp2-chost-byos-v20210304', 'environment': 'PublicAzure'}
WARNING:8c5fc3cd9b18_microsoftimages_unique_constraint.py:    {'name': 'suse-sles-15-sp2-chost-byos-v20210304-gen2', 'environment': 'Blackforest'}
DEBUG:__main__:(psycopg2.errors.UniqueViolation) could not create unique index "microsoftimages_name_environment_key"
DETAIL:  Key (name, environment)=(suse-sles-15-sp2-chost-byos-v20210304-gen2, PublicAzure) is duplicated.
...
```

Is this acceptable @bear454 ? I can have the code explicitly raise an exception rather than falling through to the existing exception if that is preferred? Or potentially do the duplicates detection and reporting in response to the exception being caught...

But that comes back to my original questions in this regard @bear454, namely how do you see this information being reported? Is the above set of warnings being reported in the logging output stream acceptable?

**Review by smarlowucf — APPROVED** [review]

**Review by bear454 — APPROVED** [review]

File: pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py (hunk 1 above)
Comment (bear454) [reply pint_server/pint_db_migrate/versions/8c5fc3cd9b18_microsoftimages_unique_constraint.py]: That's fine.
